### PR TITLE
Feature/routing bundle

### PR DESCRIPTION
**Pull request by PhilVanB**
The refactored Mapbender RoutingBundle for MB v4

### CODE REVIEW

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+        },
+
+        _getRevGeocode: function(coordinate) {
+            var self = this;
+            return $.ajax({
+                type: "GET",
+                url: self.elementUrl + 'revGeocode',
+                data: {
+                    coordinate: coordinate,
+                    srsId: self.olMap.getProjectionObject().projCode
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js:703]: Reverse Geocoding fails in my test. Error is `Uncaught TypeError: self.olMap.getProjectionObject is not a function`

Should probably use `self.olMap.getView().getProjection().getCode()` (untested though)

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+                }
+            });
+        },
+
+        _getAirLineFeature: function (response, style) {
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js:708]: an airline is a company that manages airplanes :)

I'd suggest either bee line or (IMO more descriptive) straight line

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+            var self = this;
+            var points = self._getRoutingPoints();
+            var airLineFeatures = [];
+            var waypoints = response.features[0].properties.waypoints;
+            //IE doesn't support Object.assign()
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js:713]: IE is not relevant anymore, using Object.assign is fine :)

File: src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            $minutes = substr($time, 3, strlen($time));
+            return (substr($minutes, 0, 1) == '0') ? substr($minutes, 1, 1) . ' Min' : $minutes . ' Min';
+        }
+        return $time . ' Std.';
+    }
+
+    protected function formatDistance($distance): string
+    {
+        if (floatval($distance) >= 1000) {
+            return round((floatval($distance) / 1000.0), 2) . ' KM';
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php:90]: KM is in uppercase and has a leading space, m is lowercase without a space, please unify

File: src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    protected function formatTime($time): string
+    {
+        $time = gmdate('H:i', $time);
+        if (substr($time, 0, 2) == '00') {
+            $minutes = substr($time, 3, strlen($time));
+            return (substr($minutes, 0, 1) == '0') ? substr($minutes, 1, 1) . ' Min' : $minutes . ' Min';
+        }
+        return $time . ' Std.';
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php:84]: this should also be translated

File: src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        return $this->addInstructionSymbols($this->translateInstructions($steps));
+    }
+
+    protected function formatTime($time): string
+    {
+        $time = gmdate('H:i', $time);
+        if (substr($time, 0, 2) == '00') {
+            $minutes = substr($time, 3, strlen($time));
+            return (substr($minutes, 0, 1) == '0') ? substr($minutes, 1, 1) . ' Min' : $minutes . ' Min';
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php:82]: "min" should be translated

File: src/Mapbender/RoutingBundle/Element/Routing.php [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    protected ReverseGeocodingHandler $reverseGeocodingHandler;
+
+    public function __construct(RoutingHandler $routingHandler, SearchHandler $searchHandler, ReverseGeocodingHandler $reverseGeocodingHandler) {
+        $this->routingHandler = $routingHandler;
+        $this->searchHandler = $searchHandler;
+        $this->reverseGeocodingHandler = $reverseGeocodingHandler;
+    }
+
+    /**
+     * @return string
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Element/Routing.php:36]: this comments are superflous when using native return types

File: src/Mapbender/RoutingBundle/Element/Routing.php [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    {
+        return [
+            'title' => 'Routing',
+            'advancedSettings' => false,
+            'autoSubmit' => false,
+            'allowIntermediatePoints' => false,
+            'useSearch' => false,
+            'useReverseGeocoding' => false,
+            'buffer' => 0,
+            'infoText' => '{start} → {destination} </br> {length} will take {time}',
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Element/Routing.php:89]: is it possible to translate the default here? So someone with a german machine gets a german default text when first creating the element?

File: src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        $builder
+            ->add('advancedSettings', CheckboxType::class, [
+                'required' => false,
+                'label' => 'mb.routing.backend.dialog.label.advanced',
+            ])
+            ->add('autoSubmit', CheckboxType::class, [
+                'required' => false,
+                #'property_path' => '[autoSubmit]',
```
Comment (Phocacius) [inline src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php:34]: I think they can be safely removed

**Review by Phocacius — CHANGES_REQUESTED** [review]

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js (hunk 2 above)
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: This is old and unused code. I remove it.

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js (hunk 3 above)
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: This is old and unused code. I remove it.

File: src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php (hunk 9 above)
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php]: Yes, I did so

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+            }, () => {
+                Mapbender.checkTarget('mbRoutingElement');
+            });
+            this.elementUrl = Mapbender.configuration.application.urls.element + '/' + this.element.attr('id') + '/';
+            const searchDriver = this.options.searchConfig.driver;
+            this.searchConfig = this.options.searchConfig[searchDriver];
+        },
+
+        _setup: function(mbMap) {
+            const self = this;
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: Yes, done so.

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 11]
```diff
@@ -0,0 +1,764 @@
...
+        _initializeEventListeners: function() {
+            const self = this;
+            this.olMap.on('singleclick', $.proxy(this._mapClickHandler, this));
+
+            // flush points when srs is changed
+            $(document).on('mbmapsrschanged', $.proxy(this._emptyPoints, this));
+
+            // add point on click
+            $('#addPoint', this.element).click(() => {
+                self._addInputField();
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: I changed all self to this in shorthand functions

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 12]
```diff
@@ -0,0 +1,764 @@
...
+            $('.mb-routing-location-points', this.element).change(() => {
+                if (self._isInputValid()) {
+                    self._getRoute();
+                }
+            });
+        },
+
+        _isInputValid: function() {
+            let isValid = true;
+            $.each($(this).find('.mb-routing-location-points input'), (index, element) => {
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: I changed it to `this.element`.

File: src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js [hunk 13]
```diff
@@ -0,0 +1,764 @@
...
+                return routingPoints;
+            }
+        },
+
+        _transformCoordinates: function(coordinatePair, srcProj, destinationProj) {
+            return ol.proj.transform(coordinatePair, srcProj, destinationProj);
+        },
+
+        setSpinnerVisible: function(setVisible){
+            let calculateRouteBtn = $('#calculateRoute i');
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Resources/public/mapbender.element.routing.js]: I changed it to ´ $('#calculateRoute i', this.element)´

File: src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php (hunk 6 above)
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Component/RoutingDriver/RoutingDriver.php]: Done,  my comment see above

File: src/Mapbender/RoutingBundle/Element/Routing.php (hunk 7 above)
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Element/Routing.php]: I removed them.

File: src/Mapbender/RoutingBundle/MapbenderRoutingBundle.php [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Loader\XmlFileLoader;
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+/**
+ * Mapbender routing bundle
+ *
+ * @author David Patzke
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/MapbenderRoutingBundle.php]: I removed all the authors.

File: src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+                    'min' => 0,
+                ],
+                #'property_path' => '[buffer]',
+            ])
+            ->add('infoText', TextType::class, [
+                'label' => 'mb.routing.backend.dialog.label.infoText',
+                'required' => false,
+                'empty_data' => "{start} → {destination} </br> {length} will take {time}",
+            ])
+            ->add('dateTimeFormat',TextType::class, [
```
[the remark this answers is not in the document]
Comment (PhilVanB) [reply src/Mapbender/RoutingBundle/Element/Type/RoutingAdminType.php]: I removed the configuration for the time format. It's better to implement this accordingly for each routing driver within the routing-driver-services.

**Review by Phocacius — APPROVED** [review]

**Review by astroidex — APPROVED** [review]

**Review by astroidex — CHANGES_REQUESTED** [review]
Please add an entry to the changellog.md file
someting like
New element "Routing" that allows routing via OSRM (https://github.com/mapbender/mapbender/pull/1655)

**Pull request merged**